engine/runtime/src/main/java/org/enso/interpreter/instrument/UpdatesSynchronizationState.java

```diff
@@ -0,0 +1,154 @@
+package org.enso.interpreter.instrument;
```

Comment:
Unless I'm missing something it seems like this implementation can race itself. You probably want to synchronize each modification on the thing being modified.

Reply:
Good catch! I added synchronization to all methods :+1: